rfd/0126-backend-migrations.md

```diff
@@ -0,0 +1,258 @@
+		MarshalFunc:   services.MarshalServer,
+		UnmarshalFunc: services.UnmarshalServer,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	newSvc, err := generic.NewService(&generic.ServiceConfig[types.Server]{
+		Backend:       b,
+		ResourceKind:  types.KindNode,
+		BackendPrefix: backend.Key(nodesPrefixV2, namespace),
+		MarshalFunc:   func(t types.Server) ([]byte, error) {
+			return proto.Marshal(t)
+		}
+		UnmarshalFunc: func(types.Server) ([]byte, error) { return nil, nil }
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	servers, err := oldSvc.GetResources(ctx)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			return nil
+		}
+		return trace.Wrap(err)
+	}
+
+	for _, s := range servers {
+		if err := newSvc.CreateResource(ctx, s); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	return nil
+}
+```
+
+#### Migration Backend Service
+
+Now that there is a generic backend service implementation we can provide a single implementation that attempts to reads from the new key and falling back to the previous key if not found. Whenever a migration occurs for a particular resource only the backend service would need to be converted to the generic service which already implemented the fallback mechanism.
+
+We can either extend the `generic.Service` with an optional `PreviousBackendPrefix`, `MarshalFunc`, `UnmarshalFunc` or add a new `MigrationService` that has a `generic.Service` for each `BackendPrefix` and marshal function pair like the following:
+
+```go
+type MigrationService[T types.Resource] struct {
+        previous *generic.Service[T]
+        current  *generic.Service[T]
+}
+```
+
+Retrieving a resource could be implemented as so:
+
+```go
+func (s *MigrationService[T]) GetResource(ctx context.Context, name string) (T, error) {
+        t, err := s.current.GetResource(ctx, name)
+        switch {
+        case err == nil:
+                return t, nil
+        case trace.IsNotFound(err):
+                if t, err := s.previous.GetResource(ctx, name); err == nil {
+                        return t, nil
+                }
+
+                return t, trace.Wrap(err)
+        default:
+                return t, trace.Wrap(err)
+        }
+}
+```
+
+Listing across both keys can be achieved using some of the [stream helpers](https://github.com/gravitational/teleport/tree/fspmarshall/sorted-stream-helpers) that were originally included in https://github.com/gravitational/teleport/pull/18361 but were not included in the final version of that PR since nothing was consuming them yet.
+
+### Security
+
+Migrations already exist today, this RFD only proposes a way to make them deterministic and elevates visibility into migration history. Only users with the correct permissions will be able to invoke `tctl migrations apply` and in most cases the command will result in a no-op. If all migrations have already been applied nothing will be done. We will also only allow a single migration process to be in flight at any given time.
```

Comment:
Should we emit an audit event when a migration is run? (especially if it was manually triggered with tctl)